Include `file` and `line` attribute for junit report output format

### Welcome

- [X] Yes, I've searched similar [issues on GitHub](https://github.com/golangci/golangci-lint/issues) and didn't find any.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/golangci/golangci-lint?tab=coc-ov-file)

### Your feature request related to a problem? Please describe

Various tools like [action-junit-report](https://github.com/mikepenz/action-junit-report) rely on the `file` and `line` testcase attributes in order to report errors nicely inline the file. 

 `file` and `line` testcase attributse are also part of the Common JUnit XML Format:
https://github.com/testmoapp/junitxml?tab=readme-ov-file#complete-junit-xml-example

It'd be nice to include that information in golangci-lint reports as well.

### Describe the solution you'd like

Include  `file` and `line` testcase attributes like so:
```xml
<testsuites>
  <testsuite name="go/internal/pkg/set/set_test.go" tests="1" errors="0" failures="1">
    <testcase name="unused" classname="go/internal/pkg/set/set_test.go:15:6" file="go/internal/pkg/set/set_test.go" line="15">
      <failure message="go/internal/pkg/set/set_test.go:15:6: func `unused` is unused" type=""><![CDATA[: func `unused` is unused
Category: unused
File: go/internal/pkg/set/set_test.go
Line: 15
Details: func unused(t *testing.T) {]]></failure>
    </testcase>
  </testsuite>
</testsuites>
```

### Describe alternatives you've considered

Parsing the file and line number from the classname is too brittle. 

### Additional context

_No response_

### Supporter

- [ ] I am a sponsor through [GitHub](https://github.com/sponsors/golangci/) or [OpenCollective](https://opencollective.com/golangci-lint)

--- a/.golangci.next.reference.yml
+++ b/.golangci.next.reference.yml
@@ -70,6 +70,7 @@ output:
   # - `checkstyle`
   # - `code-climate`
   # - `junit-xml`
+  # - `extended-junit-xml`
   # - `github-actions`
   # - `teamcity`
   # - `sarif`

--- a/.golangci.next.reference.yml
+++ b/.golangci.next.reference.yml
@@ -70,7 +70,7 @@ output:
   # - `checkstyle`
   # - `code-climate`
   # - `junit-xml`
-  # - `extended-junit-xml`
+  # - `junit-xml-extended`
   # - `github-actions`
   # - `teamcity`
   # - `sarif`

--- a/jsonschema/golangci.next.jsonschema.json
+++ b/jsonschema/golangci.next.jsonschema.json
@@ -520,6 +520,7 @@
                   "checkstyle",
                   "code-climate",
                   "junit-xml",
+                  "extended-junit-xml",
                   "github-actions",
                   "teamcity",
                   "sarif"

--- a/jsonschema/golangci.next.jsonschema.json
+++ b/jsonschema/golangci.next.jsonschema.json
@@ -520,7 +520,7 @@
                   "checkstyle",
                   "code-climate",
                   "junit-xml",
-                  "extended-junit-xml",
+                  "junit-xml-extended",
                   "github-actions",
                   "teamcity",
                   "sarif"

--- a/pkg/config/output.go
+++ b/pkg/config/output.go
@@ -17,7 +17,7 @@ const (
 	OutFormatCodeClimate       = "code-climate"
 	OutFormatHTML              = "html"
 	OutFormatJunitXML          = "junit-xml"
-	OutFormatExtendedJunitXML  = "extended-junit-xml"
+	OutFormatJunitXMLExtended  = "junit-xml-extended"
 	OutFormatGithubActions     = "github-actions" // Deprecated
 	OutFormatTeamCity          = "teamcity"
 	OutFormatSarif             = "sarif"
@@ -33,7 +33,7 @@ var AllOutputFormats = []string{
 	OutFormatCodeClimate,
 	OutFormatHTML,
 	OutFormatJunitXML,
-	OutFormatExtendedJunitXML,
+	OutFormatJunitXMLExtended,
 	OutFormatGithubActions,
 	OutFormatTeamCity,
 	OutFormatSarif,

--- a/pkg/config/output.go
+++ b/pkg/config/output.go
@@ -17,6 +17,7 @@ const (
 	OutFormatCodeClimate       = "code-climate"
 	OutFormatHTML              = "html"
 	OutFormatJunitXML          = "junit-xml"
+	OutFormatExtendedJunitXML  = "extended-junit-xml"
 	OutFormatGithubActions     = "github-actions" // Deprecated
 	OutFormatTeamCity          = "teamcity"
 	OutFormatSarif             = "sarif"
@@ -32,6 +33,7 @@ var AllOutputFormats = []string{
 	OutFormatCodeClimate,
 	OutFormatHTML,
 	OutFormatJunitXML,
+	OutFormatExtendedJunitXML,
 	OutFormatGithubActions,
 	OutFormatTeamCity,
 	OutFormatSarif,

--- a/pkg/printers/junitxml.go
+++ b/pkg/printers/junitxml.go
@@ -30,6 +30,8 @@ type testCaseXML struct {
 	Name      string     `xml:"name,attr"`
 	ClassName string     `xml:"classname,attr"`
 	Failure   failureXML `xml:"failure"`
+	File      string     `xml:"file,attr,omitempty"`
+	Line      int        `xml:"line,attr,omitempty"`
 }
 
 type failureXML struct {
@@ -39,11 +41,15 @@ type failureXML struct {
 }
 
 type JunitXML struct {
-	w io.Writer
+	extended bool
+	w        io.Writer
 }
 
-func NewJunitXML(w io.Writer) *JunitXML {
-	return &JunitXML{w: w}
+func NewJunitXML(extended bool, w io.Writer) *JunitXML {
+	return &JunitXML{
+		extended: extended,
+		w:        w,
+	}
 }
 
 func (p JunitXML) Print(issues []result.Issue) error {
@@ -66,6 +72,11 @@ func (p JunitXML) Print(issues []result.Issue) error {
 				Content: fmt.Sprintf("%s: %s\nCategory: %s\nFile: %s\nLine: %d\nDetails: %s",
 					i.Severity, i.Text, i.FromLinter, i.Pos.Filename, i.Pos.Line, strings.Join(i.SourceLines, "\n")),
 			},
+		}
+
+		if p.extended {
+			tc.File = i.Pos.Filename
+			tc.Line = i.Pos.Line
 		}
 
 		testSuite.TestCases = append(testSuite.TestCases, tc)

--- a/pkg/printers/junitxml_test.go
+++ b/pkg/printers/junitxml_test.go
@@ -42,13 +42,14 @@ func TestJunitXML_Print(t *testing.T) {
 		},
 	}
 
-	buf := new(bytes.Buffer)
-	printer := NewJunitXML(buf)
-
-	err := printer.Print(issues)
-	require.NoError(t, err)
-
-	expected := `<testsuites>
+	testCases := []struct {
+		desc     string
+		extended bool
+		expected string
+	}{
+		{
+			desc: "basic",
+			expected: `<testsuites>
   <testsuite name="path/to/filea.go" tests="1" errors="0" failures="1">
     <testcase name="linter-a" classname="path/to/filea.go:10:4">
       <failure message="path/to/filea.go:10:4: some issue" type="warning"><![CDATA[warning: some issue
@@ -69,7 +70,48 @@ Details: func foo() {
 }]]></failure>
     </testcase>
   </testsuite>
-</testsuites>`
+</testsuites>`,
+		},
+		{
+			desc:     "extended/complete",
+			extended: true,
+			expected: `<testsuites>
+  <testsuite name="path/to/filea.go" tests="1" errors="0" failures="1">
+    <testcase name="linter-a" classname="path/to/filea.go:10:4" file="path/to/filea.go" line="10">
+      <failure message="path/to/filea.go:10:4: some issue" type="warning"><![CDATA[warning: some issue
+Category: linter-a
+File: path/to/filea.go
+Line: 10
+Details: ]]></failure>
+    </testcase>
+  </testsuite>
+  <testsuite name="path/to/fileb.go" tests="1" errors="0" failures="1">
+    <testcase name="linter-b" classname="path/to/fileb.go:300:9" file="path/to/fileb.go" line="300">
+      <failure message="path/to/fileb.go:300:9: another issue" type="error"><![CDATA[error: another issue
+Category: linter-b
+File: path/to/fileb.go
+Line: 300
+Details: func foo() {
+	fmt.Println("bar")
+}]]></failure>
+    </testcase>
+  </testsuite>
+</testsuites>`,
+		},
+	}
 
-	assert.Equal(t, expected, buf.String())
+	for _, test := range testCases {
+		test := test
+		t.Run(test.desc, func(t *testing.T) {
+			t.Parallel()
+
+			buf := new(bytes.Buffer)
+			printer := NewJunitXML(test.extended, buf)
+
+			err := printer.Print(issues)
+			require.NoError(t, err)
+
+			assert.Equal(t, test.expected, buf.String())
+		})
+	}
 }

--- a/pkg/printers/junitxml_test.go
+++ b/pkg/printers/junitxml_test.go
@@ -101,7 +101,6 @@ Details: func foo() {
 	}
 
 	for _, test := range testCases {
-		test := test
 		t.Run(test.desc, func(t *testing.T) {
 			t.Parallel()
 

--- a/pkg/printers/printer.go
+++ b/pkg/printers/printer.go
@@ -131,7 +131,7 @@ func (c *Printer) createPrinter(format string, w io.Writer) (issuePrinter, error
 		p = NewHTML(w)
 	case config.OutFormatJunitXML:
 		p = NewJunitXML(false, w)
-	case config.OutFormatExtendedJunitXML:
+	case config.OutFormatJunitXMLExtended:
 		p = NewJunitXML(true, w)
 	case config.OutFormatGithubActions:
 		p = NewGitHubAction(w)

--- a/pkg/printers/printer.go
+++ b/pkg/printers/printer.go
@@ -130,7 +130,9 @@ func (c *Printer) createPrinter(format string, w io.Writer) (issuePrinter, error
 	case config.OutFormatHTML:
 		p = NewHTML(w)
 	case config.OutFormatJunitXML:
-		p = NewJunitXML(w)
+		p = NewJunitXML(false, w)
+	case config.OutFormatExtendedJunitXML:
+		p = NewJunitXML(true, w)
 	case config.OutFormatGithubActions:
 		p = NewGitHubAction(w)
 	case config.OutFormatTeamCity:

--- a/pkg/printers/printer.go
+++ b/pkg/printers/printer.go
@@ -115,7 +115,7 @@ func (c *Printer) createPrinter(format string, w io.Writer) (issuePrinter, error
 	switch format {
 	case config.OutFormatJSON:
 		p = NewJSON(c.reportData, w)
-	case config.OutFormatColoredLineNumber, config.OutFormatLineNumber:
+	case config.OutFormatLineNumber, config.OutFormatColoredLineNumber:
 		p = NewText(c.cfg.PrintIssuedLine,
 			format == config.OutFormatColoredLineNumber, c.cfg.PrintLinterName,
 			c.log.Child(logutils.DebugKeyTextPrinter), w)
@@ -129,10 +129,8 @@ func (c *Printer) createPrinter(format string, w io.Writer) (issuePrinter, error
 		p = NewCodeClimate(w)
 	case config.OutFormatHTML:
 		p = NewHTML(w)
-	case config.OutFormatJunitXML:
-		p = NewJunitXML(false, w)
-	case config.OutFormatJunitXMLExtended:
-		p = NewJunitXML(true, w)
+	case config.OutFormatJunitXML, config.OutFormatJunitXMLExtended:
+		p = NewJunitXML(format == config.OutFormatJunitXMLExtended, w)
 	case config.OutFormatGithubActions:
 		p = NewGitHubAction(w)
 	case config.OutFormatTeamCity:

